Fix bug of same-epoch stake deactivation after stake redelegation

HaoranYi opened the pull request:
#### Problem

Normally, activated stake has to wait at least one epoch after being deactivated, before it becomes deactivated and withdrawable. 

However, there is a bug in redelegation which could lead to same-epoch stake deactivation after stake redelegation. By redelegating the activated stake to another validator and deactivating the redelegated stake, the stake becomes deactivated and withdrawable in the same epoch.

Part of https://github.com/solana-labs/solana/pull/31876


#### Summary of Changes

Fix bug of same-epoch stake deactivation after stake redelegation



Fixes #
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

mvines reviewed (commented):
this PR is so much clearer now that the other refactoring landed earlier!
it looks like there's no test for the condition we're fixing here though, we'll definitely want to add one, or several, to demonstrate the fix

HaoranYi commented:
> this PR is so much clearer now that the other refactoring landed earlier! it looks like there's no test for the condition we're fixing here though, we'll definitely want to add one, or several, to demonstrate the fix

Yeah. Added.

t-nelson commented:
can we spruce up the description to indicate that this bug is restricted to the `Redelegate` operation?

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -614,14 +614,18 @@ pub fn delegate(
         meta.authorized.check(signers, StakeAuthorize::Staker)?;
-        stake.deactivate(clock.epoch)?;
-
+        let stake_history = invoke_context.get_sysvar_cache().get_stake_history()?;
+        stake.deactivate(clock.epoch, stake_flags, Some(&stake_history))?;
+        // After deactivation, need to clear DeactivationFlag.
+        // So that future activation and deactivation are not subject to the `MustFullyActivateBeforeDeactivationIsPermitted` restriction.
+        stake_flags.remove(StakeFlags::MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED);
```
any reason not to pass the `InvokeContext` one level deeper into `Stake::deactivate()`? then we can dereplicode and do the `StakeHistory` query and flag removal in one place rather than every call site

HaoranYi replied:
That will result in a dependency of `InvokeContext` which is in `solana-program` to sdk/programs/src/stake/state.rs. 

Currently that module only depends on class within `solana-sdk`. No foreign dependency outside its own library.

t-nelson replied:
ah. right... lame.  local helper, then?

HaoranYi replied:
yeah. will do.

HaoranYi replied:
Done.

t-nelson commented on sdk/program/src/stake/state.rs:
```diff
@@ -567,13 +567,31 @@ impl Stake {
+
+        // when MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED flag is set on stake_flags,
+        // deactivation is only permitted when the stake delegation activating amount is zero.
+        if stake_flags.contains(StakeFlags::MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED) {
```
i think we're actually going to need to lift the flag check up one level so it can be explicitly feature gated.  as is, if someone were able to manipulate a stake account to set the flag before feature activation, they could induce an error priority reordering and fork the cluster.

t-nelson commented on sdk/program/src/stake/state.rs:
```diff
@@ -567,13 +567,31 @@ impl Stake {
-            Err(StakeError::AlreadyDeactivated)
-        } else {
-            self.delegation.deactivation_epoch = epoch;
-            Ok(())
+            return Err(StakeError::AlreadyDeactivated);
```
nit: generally prefer to scope explicitly (especially in the native programs) rather than early return. early returns make introducing result priority reordering bugs far too easy

HaoranYi replied:
done

HaoranYi commented on cli/tests/stake.rs:
```diff
@@ -2320,3 +2320,757 @@ fn test_stake_minimum_delegation() {
 }
+
+#[test]
+fn test_stake_redelegation_then_deactivation_withdraw_not_permitted() {
```
I am not sure that I understand your comments. 

1. Inside cli/tests, it uses `TestValidator` class. Is that not the `SolanaTestValidator`?
2. The test scenario for this bug is very complex. It involves multiple stake accounts and votes accounts, and a sequence of very specific actions, i.e. delegate, redelegate, merge, deactivate, withdraw etc. It also requires stake_history mutation. I am not sure how this can be modeled on just bank itself. I think bank tests are more suitable to test individual instruction. Do you know any examples on bank test to model stake deactivation, activation and withdraw?
3. It seems like cli/stake.rs is where we are testing all other existing scenarios of stake manipulation, such as redelegation, withdraw etc. Therefore, I feel it is natural to extend those tests to cover the new behavior.

t-nelson replied:
i'm saying that the cli crate is not a test framework, that these tests belong elsewhere and implemented without the cli details

HaoranYi replied:
To me, the cli/tests looks more or less like a test framework. It has been used to test stake, vote, transfer, and other programs.  And it has very useful `test_util.rs`, which make it very convenient to model complex test scenario there. 

I don't know if we should "reinvent the wheel" to write another test framework to model the lifetime of a test validator. Or maybe I am missing something about the other test framework?
 
Do you know where the other test framework/utility is, which could be used to better model these tests?

t-nelson replied:
> To me, the cli/tests looks more or less like a test framework.

this is a function of those tests existing before `TestValidator` and us for some reason being very paranoid about cli behaving as expected. most, if not all of them should really be modernized at some point to avoid leading folks astray

> I don't know if we should "reinvent the wheel" to write another test framework to model the lifetime of a test validator.

we don't need to, that's literally what `TestValidator` was for.  the tests will be similar but won't a) live under the totally unrelated cli crate and b) have all of the superfluous cli machination in the way and confounding tests. you'll grab an `RpcClient` handle from the `TestValidator` instance, build the txes up and `rpc_client.send_and_confirm_transaction()` your way to victory. these should probably live under `programs/stake`, either a new `tests` directory or if one of existing source files seems appropriate

it should be a red flag when tests are being added to a crate that was otherwise unaffected by the change set that something isn't right organizationally

HaoranYi replied:
ok. i will take a look.

HaoranYi replied:
I pushed a commit to move the tests from cli to `program/stake`. 

But the ci sanity check is not happy about it. :(

Adding solana-test-validator to solana-stake-program results a circular dependency, i.e. 
solana-test-validator --> solana-runtime --> solana-stake-program --> solana-test-validator!

@t-nelson

t-nelson replied:
ugh... runtime doesn't need `solana-stake-program` all of those symbols should be available from sdk.  it must be using the re-exports instead of having been properly migrated.

i'll take a look tomorrow

t-nelson replied:
started in https://github.com/solana-labs/solana/pull/32635. need to figure out the best home for the remaining symbols

joncinque replied:
This is a common issue, and rather than leaning into the sdk more, we should resolve it using the same technique as the address-lookup-table program, which has a separate crate for tests.

In my opinion, the program crates like `solana-stake-program` should function like normal program crates, like `spl-token`. When you want those types and instructions, you pull in that crate, rather than relying on `solana-program` to have everything in it.

I know that the separate crate for tests is ugly, but it avoids bloating the sdk and resolves the circular dependencies.

HaoranYi replied:
It seems that moving the stake tests out of `cli` is not as straightforward as it should be.

How about keeping these 3 tests in `cli` for this PR, and moving all stake tests out of `cli` in a separated follow-up PR?

t-nelson replied:
> which has a separate crate for tests

oh thank god. i missed that this wasn't implementing the tests directory as a standalone crate. just realized i was about to get screwed by the instruction processor being globbed into `BUILTINS`. this is The Way :tm: 

> the program crates like solana-stake-program should function like normal program crates

generally agree, but there's a bunch of crap in this program crate that makes zero sense. even have comments about being external helpers with uses listed that are... not the program itself!

t-nelson replied:
> It seems that moving the stake tests out of `cli` is not as straightforward as it looks.
> 
> How about keeping these 3 tests in `cli` for this PR, and moving all stake tests out of `cli` in a separated follow-up PR after we fixes the circular dependencies issue?

take a crack at making the tests dir a crate as jon proposes first. doing the easy thing our of impatience is how the code got into the shape it is today. that followup pr rarely happens

HaoranYi replied:
ok. I will create a new test crate `solana-stake-program-tests` for those tests.

joncinque replied:
> generally agree, but there's a bunch of crap in this program crate that makes zero sense. even have comments about being external helpers with uses listed that are... not the program itself!

That's totally fair... the native program crates are probably the most problematic with dependencies. We could eventually make it a policy that *all* native stuff goes into the sdk, but I would prefer to thin out the sdk whenever possible.

Note: I'm the one who moved a lot of the stake types into the sdk in the first place while developing stake pools, because I didn't think to use `target_os` guards in `solana-stake-program`, so I bear a lot of the responsibility there :sweat:

HaoranYi replied:
Done!

joncinque commented on Cargo.toml:
```diff
@@ -88,6 +88,7 @@ members = [
     "sdk/program",
     "send-transaction-service",
     "stake-accounts",
+    "stake-program-test",
```
nit: let's put this in `programs/stake-tests` for consistency with address lookup table tests

joncinque commented on stake-program-test/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "stake-program-test"
```
nit:
```suggestion
name = "solana-stake-program-tests"
```

joncinque commented on stake-program-test/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+license = { workspace = true }
+edition = { workspace = true }
+
+[dependencies]
```
These should all be `dev-dependencies`

joncinque commented on stake-program-test/src/lib.rs:
nit: you should be able to delete this file

joncinque commented on stake-program-test/tests/stake.rs:
Although it's better factored, I don't think this addresses @t-nelson 's point about using the CLI as a test framework, since it's still using CLI commands.

Rather than CLI helpers, how about using `solana-program-test`? To make it easier, you can copy these helpers for creating vote accounts https://github.com/solana-labs/solana-program-library/blob/8f1a8c6e9c58c8710f0655d5e54d3f07bfacf003/stake-pool/single-pool/tests/helpers/mod.rs#L349 and stake accounts https://github.com/solana-labs/solana-program-library/blob/8f1a8c6e9c58c8710f0655d5e54d3f07bfacf003/stake-pool/single-pool/tests/helpers/stake.rs#L66

I know it's annoying to change the test framework, but it'll make maintenance a lot easier going forward, and you can use `warp_to_slot` on `ProgramTestContext` to simulate advancing epochs, which will make the test run much more quickly.

HaoranYi replied:
yeah. I was debating about whether to use or not to use those CLI helpers yesterday too. Basically, the `cli` functions used in the tests are just a wrapper for `rpc_clients`. We can certainly reimplement those helpers for tests. But they wouldn't make much difference except duplicated code. 

I will take a look at the program-test and see if rewrite those tests in program-tests is better.

HaoranYi replied:
Done! Rewrite the tests in `program-test`.

joncinque reviewed (commented):
We're getting really close! Just a few bits to make the tests better, which @t-nelson can shoot down if he wishes.

joncinque commented on program-test/tests/stake.rs:
```diff
@@ -0,0 +1,190 @@
+        context.last_blockhash,
+    );
+    let r = context.banks_client.process_transaction(transaction).await;
+    assert_eq!(
+        format!("{}", r.err().unwrap()),
+        "transport transaction error: Error processing Instruction 0: custom program error: 0xf"
+    );
```
Please fix up this error too using the proper typing and error code

joncinque commented on program-test/tests/stake.rs:
```diff
@@ -0,0 +1,190 @@
+        context.last_blockhash,
+    );
+    let r = context.banks_client.process_transaction(transaction).await;
+    assert!(format!("{}", r.err().unwrap()).contains("insufficient funds for instruction"));
```
Same here

joncinque reviewed (commented):
The test looks great! Mainly some nits around it, so this is *extremely* close

joncinque reviewed (commented):
Looks `setup.rs` wasn't committed, but the rest of the changes look good

HaoranYi commented:
> Looks `setup.rs` wasn't committed, but the rest of the changes look good

oops. added.

joncinque reviewed (approved):
Looks great! Be sure to get at least one more approval on this since it touches the stake program

t-nelson commented on programs/stake/src/stake_instruction.rs:
```diff
@@ -6458,6 +6459,10 @@ mod tests {
                                 ..Clock::default()
                             }),
                         ),
+                        (
```
it's not clear what this addition is for. don't we get stake history from the invoke context / sysvar cache?

HaoranYi replied:
This to mock StakeHistory account for test. deactivate fn now requires StakeHistory to be available. invoke context/sysvar cache will only have the StakeHistory in the cache when StakeHistory account exists. Therefore, StakeHistory account need to be mocked for the test.

CriesofCarrots replied:
Can confirm that this is necessary for our test framework

t-nelson commented on program-test/tests/warp.rs:
```diff
@@ -209,11 +134,30 @@ async fn clock_sysvar_updated_from_warp() {
     let program_test = ProgramTest::default();
     let mut context = program_test.start_with_context().await;
+
+    context.warp_to_slot(100).unwrap();
```
should _really_ avoid mixing functional and cosmetic changes like this. it makes review very error prone. i spent 10min trying to figure out what this addition was for before noticing that it was pulled out of `setup_vote()`

HaoranYi replied:
yeah. 
Need to move wrap out so that we can call setup mutiple vote accounts. To keep the original caller to have the same behavior, those "warp" are inserted.

CriesofCarrots replied:
To be clear, Trent isn't objecting to the change, just requesting that the move and the edit be noted in separate commits. So the commits in this PR might be something like:
1. Add `deactivate_stake()` helper
2. Return error when feature activated and StakeFlag set
3. Add stake_flag unit test
4. Move `setup_stake` to its own mod
5. Only warp in warp.rs tests
6. Add stake.rs tests

I don't want to bog this down in a reorg, so just a consideration for next time.

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -613,15 +613,48 @@ pub fn delegate(
+            .delegation
+            .stake_activating_and_deactivating(epoch, Some(&stake_history));
+        if status.activating != 0 {
+            Err(StakeError::RedelegatedStakeMustFullyActivateBeforeDeactivationIsPermitted.into())
```
```suggestion
            Err(InstructionError::from(StakeError::RedelegatedStakeMustFullyActivateBeforeDeactivationIsPermitted))
```

`<T>::from(x)` self-documents better than `x.into()`

HaoranYi replied:
Done.

t-nelson commented on programs/stake/src/stake_state.rs:
```diff
@@ -613,15 +613,48 @@ pub fn delegate(
+    if invoke_context
+        .feature_set
+        .is_active(&feature_set::stake_redelegate_instruction::id())
+        && stake_flags.contains(StakeFlags::MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED)
```
prefer not combining feature-gate activation test with feature-gated logic to simplify feature-gate removal

HaoranYi replied:
Done.

t-nelson reviewed (commented):
i'm mostly down to nits :partying_face:

HaoranYi commented:
I rebase this PR to the master and resolve the conflicts. 

Specifically, there are two more changes since last review:
1. deprecate `StakeState`, use `StakeStateWithFlags`.
2. add warmup/cooldown rate epoch to pending stake amount calculation during redelegate check. 

@t-nelson @joncinque Please review again. Thanks!

joncinque commented on program-test/tests/warp.rs:
```diff
@@ -29,89 +30,13 @@ use {
-        vote_instruction,
-        vote_state::{self, VoteInit, VoteState},
-    },
+    solana_vote_program::vote_state::{self},
```
micro-nit:
```suggestion
    solana_vote_program::vote_state,
```

joncinque replied:
It looks like this nit didn't make it in

HaoranYi replied:
updated.

joncinque reviewed (dismissed):
Looks good to me! Just the micro-nit

HaoranYi commented:
@t-nelson Can you review this again and let me know if you are okay with this PR? Thanks!

joncinque commented on program-test/tests/stake.rs:
```diff
@@ -0,0 +1,203 @@
+        .unwrap();
+}
+
+#[tokio::test]
+async fn test_stake_redelegation_then_deactivation_withdraw_not_permitted() {
+    test_stake_redelegation_pending_activation(PendingStakeActivationTestFlag::NoMerge).await
+}
+
+#[tokio::test]
+async fn test_stake_redelegation_then_merge_active_then_deactivation_withdraw_not_permitted() {
+    test_stake_redelegation_pending_activation(PendingStakeActivationTestFlag::MergeActive).await
+}
+
+#[tokio::test]
+async fn test_stake_redelegation_then_merge_inactive_then_deactivation_withdraw_not_permitted() {
+    test_stake_redelegation_pending_activation(PendingStakeActivationTestFlag::MergeInactive).await
+}
```
nit: rather than having all of these, you could have them all as separate `test_case`s

HaoranYi replied:
good idea. updated.

joncinque reviewed (approved):
Still looks good to me!

CriesofCarrots commented on programs/stake/src/stake_state.rs:
```diff
@@ -633,15 +633,57 @@ pub fn delegate(
+    stake_flags: &mut StakeFlags,
+    epoch: Epoch,
+) -> Result<(), InstructionError> {
+    let stake_history = invoke_context.get_sysvar_cache().get_stake_history()?;
```
Can this be moved inside the `if` cases so it only affects the flagged accounts? No need to load this big sysvar if not needed.

HaoranYi replied:
done.

CriesofCarrots commented on program-test/tests/stake.rs:
```diff
@@ -0,0 +1,193 @@
+    // 3. create 2nd vote account
+    let vote_address2 = setup_vote(&mut context).await;
+
+    // 3.1 relegate 1st stake account to 2nd stake account.
```
```suggestion
    // 3.1 relegate stake account to 2nd vote account, which creates 2nd stake account
```

HaoranYi replied:
done.

CriesofCarrots reviewed (commented):
One suggestion to move sysvar load
